Missing Proguard/R8 rules with `moshi-sealed-codegen`

`moshi-sealed-codegen` does not seem to work as advertised with minification enabled.

Reproducer: https://github.com/mhelder/moshix-sealed-missing-proguard-rules
// edit: forgot to push my branch 🤦 - fixed now!

It takes the `Message` sample from the [Moshi-sealed docs](https://github.com/ZacSweers/MoshiX/tree/main/moshi-sealed#moshi-sealed) and does this:

```kotlin
val json = """{ "type": "success", "value": "Yay" }"""
val moshi = Moshi.Builder().build()
val adapter = moshi.adapter(Message::class.java)
val message = adapter.fromJson(json)
```

Works fine with `isMinifyEnabled = false`. It'll output `Success(value=Yay)` on screen.

Now flip the `isMinifyEnabled` flag  for the `debug` build type or run the `release`.

R8 full mode on (default):
```
FATAL EXCEPTION: main
Process: com.example.app, PID: 19895
java.lang.RuntimeException: Unable to start activity ComponentInfo{com.example.app/com.example.app.MainActivity}: java.lang.IllegalArgumentException: Cannot serialize abstract class com.example.app.data.model.Message
	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:4164)
	at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:4322)
	at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:103)
	at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:139)
	at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:96)
	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2685)
	at android.os.Handler.dispatchMessage(Handler.java:106)
	at android.os.Looper.loopOnce(Looper.java:230)
	at android.os.Looper.loop(Looper.java:319)
	at android.app.ActivityThread.main(ActivityThread.java:8893)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:608)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1103)
Caused by: java.lang.IllegalArgumentException: Cannot serialize abstract class com.example.app.data.model.Message
	at com.squareup.moshi.ClassJsonAdapter$1.create(ClassJsonAdapter.java:93)
	at com.squareup.moshi.Moshi.adapter(Moshi.java:146)
	at com.squareup.moshi.Moshi.adapter(Moshi.java:106)
	at com.squareup.moshi.Moshi.adapter(Moshi.java:80)
	at com.example.app.MainActivity.onCreate(MainActivity.kt:16)
	at android.app.Activity.performCreate(Activity.java:8944)
	at android.app.Activity.performCreate(Activity.java:8913)
	at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1456)
	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:4146)
	at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:4322) 
	at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:103) 
	at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:139) 
	at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:96) 
	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2685) 
	at android.os.Handler.dispatchMessage(Handler.java:106) 
	at android.os.Looper.loopOnce(Looper.java:230) 
	at android.os.Looper.loop(Looper.java:319) 
	at android.app.ActivityThread.main(ActivityThread.java:8893) 
	at java.lang.reflect.Method.invoke(Native Method) 
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:608) 
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1103) 
```

R8 full mode off (set `android.enableR8.fullMode=false` in `gradle.properties`):
```
FATAL EXCEPTION: main
Process: com.example.app, PID: 20057
java.lang.RuntimeException: Unable to start activity ComponentInfo{com.example.app/com.example.app.MainActivity}: java.lang.RuntimeException: Failed to find the generated JsonAdapter class for class com.example.app.data.model.Message
    at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:4164)
    at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:4322)
    at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:103)
    at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:139)
    at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:96)
    at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2685)
    at android.os.Handler.dispatchMessage(Handler.java:106)
    at android.os.Looper.loopOnce(Looper.java:230)
    at android.os.Looper.loop(Looper.java:319)
    at android.app.ActivityThread.main(ActivityThread.java:8893)
    at java.lang.reflect.Method.invoke(Native Method)
    at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:608)
    at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1103)
Caused by: java.lang.RuntimeException: Failed to find the generated JsonAdapter class for class com.example.app.data.model.Message
    at com.squareup.moshi.internal.Util.generatedAdapter(Util.java:590)
    at com.squareup.moshi.StandardJsonAdapters$1.create(StandardJsonAdapters.java:61)
    at com.squareup.moshi.Moshi.adapter(Moshi.java:146)
    at com.squareup.moshi.Moshi.adapter(Moshi.java:106)
    at com.squareup.moshi.Moshi.adapter(Moshi.java:80)
    at com.example.app.MainActivity.onCreate(MainActivity.kt:16)
    at android.app.Activity.performCreate(Activity.java:8944)
    at android.app.Activity.performCreate(Activity.java:8913)
    at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1456)
    at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:4146)
    at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:4322) 
    at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:103) 
    at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:139) 
    at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:96) 
    at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2685) 
    at android.os.Handler.dispatchMessage(Handler.java:106) 
    at android.os.Looper.loopOnce(Looper.java:230) 
    at android.os.Looper.loop(Looper.java:319) 
    at android.app.ActivityThread.main(ActivityThread.java:8893) 
    at java.lang.reflect.Method.invoke(Native Method) 
    at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:608) 
    at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1103) 
Caused by: java.lang.ClassNotFoundException: com.example.app.data.model.MessageJsonAdapter
    at java.lang.Class.classForName(Native Method)
    at java.lang.Class.forName(Class.java:536)
    at com.squareup.moshi.internal.Util.generatedAdapter(Util.java:564)
    at com.squareup.moshi.StandardJsonAdapters$1.create(StandardJsonAdapters.java:61) 
    at com.squareup.moshi.Moshi.adapter(Moshi.java:146) 
    at com.squareup.moshi.Moshi.adapter(Moshi.java:106) 
    at com.squareup.moshi.Moshi.adapter(Moshi.java:80) 
    at com.example.app.MainActivity.onCreate(MainActivity.kt:16) 
    at android.app.Activity.performCreate(Activity.java:8944) 
    at android.app.Activity.performCreate(Activity.java:8913) 
    at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1456) 
    at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:4146) 
    at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:4322) 
    at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:103) 
    at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:139) 
    at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:96) 
    at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2685) 
    at android.os.Handler.dispatchMessage(Handler.java:106) 
    at android.os.Looper.loopOnce(Looper.java:230) 
    at android.os.Looper.loop(Looper.java:319) 
    at android.app.ActivityThread.main(ActivityThread.java:8893) 
    at java.lang.reflect.Method.invoke(Native Method) 
    at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:608) 
    at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1103) 
Caused by: java.lang.ClassNotFoundException: Didn't find class "com.example.app.data.model.MessageJsonAdapter" on path: DexPathList[[dex file "/data/data/com.example.app/code_cache/.overlay/base.apk/classes.dex", zip file "/data/app/~~wmAybjC_Rs2XrETNE2mbJg==/com.example.app-3vp2orfM7M1VnaqJ66cISw==/base.apk"],nativeLibraryDirectories=[/data/app/~~wmAybjC_Rs2XrETNE2mbJg==/com.example.app-3vp2orfM7M1VnaqJ66cISw==/lib/arm64, /system/lib64, /system/system_ext/lib64]]
    at dalvik.system.BaseDexClassLoader.findClass(BaseDexClassLoader.java:259)
    at java.lang.ClassLoader.loadClass(ClassLoader.java:379)
    at java.lang.ClassLoader.loadClass(ClassLoader.java:312)
    at java.lang.Class.classForName(Native Method) 
    at java.lang.Class.forName(Class.java:536) 
    at com.squareup.moshi.internal.Util.generatedAdapter(Util.java:564) 
    at com.squareup.moshi.StandardJsonAdapters$1.create(StandardJsonAdapters.java:61) 
    at com.squareup.moshi.Moshi.adapter(Moshi.java:146) 
    at com.squareup.moshi.Moshi.adapter(Moshi.java:106) 
    at com.squareup.moshi.Moshi.adapter(Moshi.java:80) 
    at com.example.app.MainActivity.onCreate(MainActivity.kt:16) 
    at android.app.Activity.performCreate(Activity.java:8944) 
    at android.app.Activity.performCreate(Activity.java:8913) 
    at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1456) 
    at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:4146) 
    at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:4322) 
    at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:103) 
    at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:139) 
    at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:96) 
    at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2685) 
    at android.os.Handler.dispatchMessage(Handler.java:106) 
    at android.os.Looper.loopOnce(Looper.java:230) 
    at android.os.Looper.loop(Looper.java:319) 
    at android.app.ActivityThread.main(ActivityThread.java:8893) 
    at java.lang.reflect.Method.invoke(Native Method) 
    at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:608) 
    at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1103) 
```

Project has no Proguard/R8 rules defined.

Rules generated under `resources/META-INF/proguard`:
- **moshi-com.example.app.data.model.Message.Success.pro**:
```
-if class com.example.app.data.model.Message$Success
-keepnames class com.example.app.data.model.Message$Success
-if class com.example.app.data.model.Message$Success
-keep class com.example.app.data.model.Message_SuccessJsonAdapter {
    public <init>(com.squareup.moshi.Moshi);
}
```

- **moshi-com.example.app.data.model.Message.pro**:
```
<empty file>
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -8,6 +8,7 @@ Changelog
 - Fix fallback adapter support in IR code gen not recognizing Moshi parameters to primary constructors.
 - Check for same subtypes before erroring on duplicate labels in moshi-sealed IR.
 - Fix proguard rule gen not capturing non-sealed subtypes.
+- Don't write empty proguard rule files if not rules were necessary.
 
 0.27.0
 ------

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -7,6 +7,7 @@ Changelog
 - Improve moshi-sealed KSP error messages.
 - Fix fallback adapter support in IR code gen not recognizing Moshi parameters to primary constructors.
 - Check for same subtypes before erroring on duplicate labels in moshi-sealed IR.
+- Fix proguard rule gen not capturing non-sealed subtypes.
 
 0.27.0
 ------

--- a/moshi-proguard-rule-gen/src/main/kotlin/dev/zacsweers/moshix/proguardgen/MoshiProguardGenSymbolProcessor.kt
+++ b/moshi-proguard-rule-gen/src/main/kotlin/dev/zacsweers/moshix/proguardgen/MoshiProguardGenSymbolProcessor.kt
@@ -163,6 +163,8 @@ public class MoshiProguardGenSymbolProcessor(private val environment: SymbolProc
       for (nested in getSealedSubclasses()) {
         nested.walkSealedSubtypes(elements, skipAnnotationCheck = false)
       }
+    } else {
+      elements += toClassName()
     }
   }
 

--- a/moshi-proguard-rule-gen/src/main/kotlin/dev/zacsweers/moshix/proguardgen/MoshiProguardGenSymbolProcessor.kt
+++ b/moshi-proguard-rule-gen/src/main/kotlin/dev/zacsweers/moshix/proguardgen/MoshiProguardGenSymbolProcessor.kt
@@ -98,6 +98,11 @@ public class MoshiProguardGenSymbolProcessor(private val environment: SymbolProc
               targetConstructorHasDefaults = false,
               targetConstructorParams = emptyList(),
             )
+
+          if (!enableMoshi && nestedSealedClassNames.isEmpty()) {
+            return emptyList()
+          }
+
           environment.logger.info(
             "MOSHI: Writing proguard rules for ${targetType.canonicalName}: $config",
             clazz,

--- a/moshi-proguard-rule-gen/src/test/kotlin/dev/zacsweers/moshix/proguardgen/MoshiProguardGenSymbolProcessorTest.kt
+++ b/moshi-proguard-rule-gen/src/test/kotlin/dev/zacsweers/moshix/proguardgen/MoshiProguardGenSymbolProcessorTest.kt
@@ -97,6 +97,65 @@ class MoshiProguardGenSymbolProcessorTest(private val useKSP2: Boolean) {
   }
 
   @Test
+  fun ensureNestedSealedClassesAreAdded() {
+    val source =
+      kotlin(
+        "BaseType.kt",
+        """
+      package test
+      import com.squareup.moshi.Json
+      import com.squareup.moshi.JsonClass
+      import dev.zacsweers.moshix.sealed.annotations.TypeLabel
+      import dev.zacsweers.moshix.sealed.annotations.NestedSealed
+      import dev.zacsweers.moshix.sealed.annotations.DefaultObject
+
+      @JsonClass(generateAdapter = true, generator = "sealed:type")
+        sealed class Message {
+
+            @TypeLabel("success")
+            @JsonClass(generateAdapter = true)
+            data class Success(
+                @Json(name = "value")
+                val value: String
+            ) : Message()
+
+            @DefaultObject
+            object Unknown : Message()
+        }
+    """,
+      )
+
+    val compilation = prepareCompilation(source)
+    val result = compilation.compile()
+    assertThat(result.exitCode).isEqualTo(ExitCode.OK)
+    val generatedSourcesDir = compilation.kspSourcesDir
+    val proguardFiles = generatedSourcesDir.walkTopDown().filter { it.extension == "pro" }.toList()
+    check(proguardFiles.isNotEmpty())
+    for (generatedFile in proguardFiles) {
+      generatedFile.assertInCorrectPath()
+      when (generatedFile.nameWithoutExtension) {
+        "moshi-test.Message" ->
+          assertThat(generatedFile.readText().trimIndent())
+            .isEqualTo(
+              """
+                  # Conditionally keep this adapter for every possible nested subtype that uses it.
+                  -if class test.Message.Success
+                  -keep class test.MessageJsonAdapter {
+                      public <init>(com.squareup.moshi.Moshi);
+                  }
+                  -if class test.Message.Unknown
+                  -keep class test.MessageJsonAdapter {
+                      public <init>(com.squareup.moshi.Moshi);
+                  }
+                """
+                .trimIndent()
+            )
+        else -> error("Unrecognized proguard file: $generatedFile")
+      }
+    }
+  }
+
+  @Test
   fun disableProguardGeneration() {
     val source =
       kotlin(

--- a/moshi-proguard-rule-gen/src/test/kotlin/dev/zacsweers/moshix/proguardgen/MoshiProguardGenSymbolProcessorTest.kt
+++ b/moshi-proguard-rule-gen/src/test/kotlin/dev/zacsweers/moshix/proguardgen/MoshiProguardGenSymbolProcessorTest.kt
@@ -80,11 +80,23 @@ class MoshiProguardGenSymbolProcessorTest(private val useKSP2: Boolean) {
       generatedFile.assertInCorrectPath()
       when (generatedFile.nameWithoutExtension) {
         "moshi-test.BaseType" ->
-          assertThat(generatedFile.readText())
-            .contains(
+          assertThat(generatedFile.readText().trimIndent())
+            .isEqualTo(
               """
                   # Conditionally keep this adapter for every possible nested subtype that uses it.
+                  -if class test.BaseType.TypeA
+                  -keep class test.BaseTypeJsonAdapter {
+                      public <init>(com.squareup.moshi.Moshi);
+                  }
+                  -if class test.BaseType.TypeB
+                  -keep class test.BaseTypeJsonAdapter {
+                      public <init>(com.squareup.moshi.Moshi);
+                  }
                   -if class test.BaseType.TypeC
+                  -keep class test.BaseTypeJsonAdapter {
+                      public <init>(com.squareup.moshi.Moshi);
+                  }
+                  -if class test.BaseType.TypeC.TypeCImpl
                   -keep class test.BaseTypeJsonAdapter {
                       public <init>(com.squareup.moshi.Moshi);
                   }
@@ -222,8 +234,8 @@ sealed class BaseType {
       generatedFile.assertInCorrectPath()
       when (generatedFile.nameWithoutExtension) {
         "moshi-test.BaseType" ->
-          assertThat(generatedFile.readText())
-            .contains(
+          assertThat(generatedFile.readText().trimIndent())
+            .isEqualTo(
               """
                   -if class test.BaseType
                   -keepnames class test.BaseType
@@ -233,7 +245,19 @@ sealed class BaseType {
                   }
 
                   # Conditionally keep this adapter for every possible nested subtype that uses it.
+                  -if class test.BaseType.TypeA
+                  -keep class test.BaseTypeJsonAdapter {
+                      public <init>(com.squareup.moshi.Moshi);
+                  }
+                  -if class test.BaseType.TypeB
+                  -keep class test.BaseTypeJsonAdapter {
+                      public <init>(com.squareup.moshi.Moshi);
+                  }
                   -if class test.BaseType.TypeC
+                  -keep class test.BaseTypeJsonAdapter {
+                      public <init>(com.squareup.moshi.Moshi);
+                  }
+                  -if class test.BaseType.TypeC.TypeCImpl
                   -keep class test.BaseTypeJsonAdapter {
                       public <init>(com.squareup.moshi.Moshi);
                   }

